prow/crier/controller.go

```diff
@@ -0,0 +1,183 @@
+limitations under the License.
+*/
+
+// package crier reports finished prowjob status to git providers
+package crier
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	pjinformers "k8s.io/test-infra/prow/client/informers/externalversions/prowjobs/v1"
+)
+
+// Controller struct defines how a controller should encapsulate
+// logging, client connectivity, informing (list and watching)
+// queueing, and handling of resource changes
+type Controller struct {
+	clientset kubernetes.Interface
+	queue     workqueue.RateLimitingInterface
+	informer  pjinformers.ProwJobInformer
+}
+
+func NewController(clientset kubernetes.Interface, queue workqueue.RateLimitingInterface, informer pjinformers.ProwJobInformer) *Controller {
+	logrus.Info("NewController: Creating")
+
+	return &Controller{
+		clientset: clientset,
+		queue:     queue,
+		informer:  informer,
+	}
+}
+
+// Run is the main path of execution for the controller loop
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	// handle a panic with logging and exiting
+	defer utilruntime.HandleCrash()
+	// ignore new items in the queue but when all goroutines
+	// have completed existing items then shutdown
+	defer c.queue.ShutDown()
+
+	logrus.Info("Controller.Run: initiating")
+	c.informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			logrus.Infof("Add prowjob: %s", key)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(newObj)
+			logrus.Infof("Update prowjob: %s", key)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, _ := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			logrus.Infof("[NOOP] Delete prowjob: %s", key)
+		},
+	})
+
+	// run the informer to start listing and watching resources
+	go c.informer.Informer().Run(stopCh)
+
+	// do the initial synchronization (one time) to populate resources
+	if !cache.WaitForCacheSync(stopCh, c.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Error syncing cache"))
+		return
+	}
+	logrus.Info("Controller.Run: cache sync complete")
+
+	// run the runWorker method every second with a stop channel
+	wait.Until(c.runWorker, time.Second, stopCh)
+}
+
+// HasSynced allows us to satisfy the Controller interface
+// by wiring up the informer's HasSynced method to it
+func (c *Controller) HasSynced() bool {
+	return c.informer.Informer().HasSynced()
+}
+
+// runWorker executes the loop to process new items added to the queue
+func (c *Controller) runWorker() {
+	logrus.Info("Controller.runWorker: starting")
```

Comment:
These seem a little superfluous and will spam very heavily at `Info` -- might want to just remove all of the alerts that aren't embedding data